GH-132657: Add lock-free set contains implementation

nascheme opened the pull request:
This roughly follows what was done for dictobject to make a lock-free lookup operation.  The set contains operation scales much better when used from multiple-threads.  The frozenset contains performance seems unchanged (as already lock-free).

Summary of changes:

- refactor `set_lookkey()` into `set_do_lookup()` which now takes a function pointer that does the entry comparison.  This is similar to dictobject and `do_lookup()`.  In an optimized build, the comparison function  is inlined and there should be no performance cost to this.
- change `set_do_lookup` to return a status separately from the entry value.  
- add `set_compare_frozenset()` and use if the object is a frozenset.  For the free-threaded build, this avoids some overhead (locking, atomic operations, incref/decref on key)
- use `FT_ATOMIC_*` macros as needed for atomic loads and stores
- use a deferred free on the set table array, if shared (only on free-threaded build, normal build always does an immediate free)
- for free-threaded build, use explicit for loop to zero the table, rather than `memcpy()`.  
- when mutating the set, assign `so->table` to NULL while the change is a happening.   Assign the real table array after the change is done.


Free-threading scaling benchmark results from the attached scripts (result for 8 cores in parallel).  This is a modified version of the `ftscalingbenchmark.py` script.


|                           | base        | this PR     |
| ------------------------- | ----------- | ----------- |
| dict_contains             | 9.1x faster | 9.4x faster |
| tuple_contains            | 8.0x faster | 7.6x faster |
| list_contains             | 7.9x faster | 7.8x faster |
| frozenset_contains        | 7.8x faster | 7.8x faster |
| frozenset_contains_dunder | 7.6x faster | 7.7x faster |
| set_contains              | 2.6x slower | 7.7x faster |
| set_contains_dunder       | 2.8x slower | 6.1x faster |
| shallow_copy              | 1.8x faster | 1.8x faster |
| deepcopy                  | 4.2x faster | 4.5x faster |

[ftscaling_set.py.txt](https://github.com/user-attachments/files/21197164/ftscaling_set.py.txt)

Separate repo containing multi-threading stress tests:
https://github.com/nascheme/python_set_stress

<!-- gh-issue-number: gh-132657 -->
* Issue: gh-132657
<!-- /gh-issue-number -->

nascheme committed:
wip: lock-free set contains

kumaraditya303 commented on Objects/setobject.c:
```diff
@@ -62,6 +62,152 @@ static PyObject _dummy_struct;
+
+#ifdef Py_GIL_DISABLED
+
+#define LOAD_HASH(entry) _Py_atomic_load_ssize_acquire(&(entry)->hash)
```
I think LOAD_HASH and STORE_HASH can use relaxed ordering if we make sure they are stored before STORE_KEY with release order so that it would synchronize with LOAD_KEY with acquire ordering.

nascheme replied:
Yeah, that seems to be okay.  At least, TSAN doesn't warn about it.  I've switched the code to directly use the `FT_ATOMIC_*` macros since I think that's clearer (even if it's more verbose).

nascheme committed:
Use FT_ATOMIC_* macros.

This makes for longer code vs using the custom LOAD_*/STORE_* macros.
However, I think this makes the code more clear.

nascheme committed:
Increase items and loops for set test.

nascheme committed:
Re-order some atomic store operations.

nascheme committed:
Add and use set_compare_frozenset().

nascheme committed:
Merge 'origin/main' into set_lockfree_contains

nascheme committed:
Fix _PyMem_FreeDelayed() calls, need size.

nascheme committed:
Fix frozenset contains method.

nascheme committed:
Add NEWS.

nascheme committed:
Re-generate clinic output.

eendebakpt commented on Misc/NEWS.d/next/Core_and_Builtins/2025-07-11-19-57-27.gh-issue-132657.vwDuO2.rst:
```diff
@@ -0,0 +1,2 @@
+For the free-threaded build, avoid locking the set object for the
```
```suggestion
For the free-threaded build, avoid locking the :class:`set` object for the
```

eendebakpt commented on Objects/setobject.c:
```diff
@@ -63,6 +63,149 @@ static PyObject _dummy_struct;
+    }
+    Py_ssize_t ep_hash = ep->hash;
+    if (ep_hash == hash) {
+        if (PyUnicode_CheckExact(startkey)
```
This optimization was introduced to avoid the check on mutating tables and the incref/decref on the `startkey`. Since that is not relevant for the frozenset, we can could perhaps remove this fast path. (there will still be a minor gain because unicode_eq is used directly, but the `PyUnicode_CheckExact`  check also takes time for the non-unicode cases).

See https://github.com/eendebakpt/cpython/commit/93035c44fd49359e47947fa23aded502d3a4155e, text `Hacked up version...`

nascheme replied:
Okay.  Based on my benchmarking, the difference is small.  So, removing that special case seems better.

nascheme committed:
Better markup in NEWS.

nascheme committed:
Remove unicode case for set_compare_frozenset.

nascheme committed:
Add multi-threaded frozenset contains test.

nascheme committed:
Merge 'origin/main' into set_lockfree_contains

nascheme committed:
Remove unused line of code, from merge.

eendebakpt commented on Lib/test/test_free_threading/test_set.py:
```diff
@@ -36,6 +36,70 @@ def repr_set():
+        NUM_LOOPS = 20
+
+        s = frozenset()
+        def make_set():
```
Why is the `make_set` here in a thread? It just assigns the same frozenset to `s` each time.

Maybe change `make_set` so that it creates a different `frozenset` each time?

nascheme replied:
I don't understand this comment.  The `make_set()` binds `s` to a new frozenset object (at least I intend it to).

eendebakpt replied:
The `s` is indeed updated with a new `frozenset`, but it is a `frozenset` with the same elements each time. I was thinking about something like
```
def make_set():
            nonlocal s
            barrier.wait()
            num_elements = NUM_ITEMS
            while not done:
               num_elements = (num_elements  + 17 %) NUM_ITEMS
                s = frozenset(range(num_elements))
```
But what are we testing here. There is no concurrent mutation of the `s`, so this test is not specific to the free-threading build I think?

nascheme replied:
Oh yes, I'm not sure what I was thinking when writing that test.  It doesn't test what I intended.  I re-wrote it so that a `set()` is used as the key and it mutated.

eendebakpt commented on Objects/setobject.c:
```diff
@@ -63,6 +63,147 @@ static PyObject _dummy_struct;
+#define SET_MARK_SHARED(so) _PyObject_GC_SET_SHARED(so)
+
+static void
+ensure_shared_on_read(PySetObject *so)
```
Note to self: this is identical to the `ensure_shared_on_read` from `dictobject.c`. The `SET_DICT_SHARED` would correspond to `SET_SET_SHARED`, which is an odd name so here it is `SET_MARK_SHARED`.

eendebakpt commented on Objects/setobject.c:
```diff
@@ -1277,6 +1474,14 @@ set_swap_bodies(PySetObject *a, PySetObject *b)
         FT_ATOMIC_STORE_SSIZE_RELAXED(a->hash, -1);
         FT_ATOMIC_STORE_SSIZE_RELAXED(b->hash, -1);
     }
+    if (!SET_IS_SHARED(b) && SET_IS_SHARED(a)) {
```
`set_swap_bodies` is only called in two locations in this file. For both cases the second argument `b` is a newly created temporary which is discarded afterwards. So we could replace part (all?) of these branches with asserts.  Part of the other work (e.g. copying back to `b`) might not be needed for the same reason.

nascheme replied:
That would simplify the code and be slightly faster (not sure how much it actually matters).  I'll create a separate issue for that so we can benchmark it.

eendebakpt commented on Objects/setobject.c:
```diff
@@ -2222,39 +2427,24 @@ set_add_impl(PySetObject *so, PyObject *key)
-        Py_BEGIN_CRITICAL_SECTION(so);
-        rv = set_contains_lock_held(so, key);
-        Py_END_CRITICAL_SECTION();
+        return set_contains_entry(so, key, hash);
```
```suggestion
```

eendebakpt commented on Objects/setobject.c:
```diff
@@ -282,13 +399,78 @@ set_insert_clean(setentry *table, size_t mask, PyObject *key, Py_hash_t hash)
+{
+    int status;
+    setentry *entry;
+    if (PyFrozenSet_CheckExact(so)) {
```
Note: this path can be taken. `set_lookkey_threadsafe` is only called in set_contains_entry. The `set_contains_entry` is only called in `_PySet_Contains`. `_PySet_Contains` is called in `set___contains___impl` and in `bytecodes.c _CONTAINS_OP_SET` (the latter can have frozenset as input)

eendebakpt commented on Objects/setobject.c:
```diff
@@ -2301,12 +2490,23 @@ static PyObject *
-    if (result < 0)
+    Py_hash_t hash = _PyObject_HashFast(key);
+    if (hash == -1) {
+        if (!PySet_Check(key) || !PyErr_ExceptionMatches(PyExc_TypeError)) {
+            set_unhashable_type(key);
+            return NULL;
+        }
+        PyErr_Clear();
+        Py_BEGIN_CRITICAL_SECTION(key);
+        hash = frozenset_hash_impl(key);
+        Py_END_CRITICAL_SECTION();
```
Note: this code looks a bit odd: why add a critical section around a call to `frozenset_hash_impl`? The reason is that `key` can be both a `set` and `frozenset`. The `frozenset_hash_impl` handles the `set`, but to make it thread safe we add a critical section. The code is the same in current main (but moved around a bit)

In python testing for membership in a `set` with an object of type `set` is allowed, even though a `set` is not hashable. E.g. `set() in {frozenset(), frozenset([1,2])}` returns `True`. The same does not hold for other container types, e.g. `set() in {frozenset(): 2}` returns a `TypeError`.

@nascheme Maybe add a comment to clarify?

nascheme replied:
Comment added.

eendebakpt commented on Objects/setobject.c:
```diff
@@ -63,6 +63,147 @@ static PyObject _dummy_struct;
+    }
+    Py_ssize_t ep_hash = FT_ATOMIC_LOAD_SSIZE_ACQUIRE(ep->hash);
+    if (ep_hash == hash) {
+        if (startkey == NULL || !_Py_TryIncrefCompare(&ep->key, startkey)) {
```
The `startkey` can never by NULL here because of the check 9 lines above.

eendebakpt commented on Objects/setobject.c:
```diff
@@ -63,6 +63,147 @@ static PyObject _dummy_struct;
+        if (cmp < 0) {
+            return SET_LOOKKEY_ERROR;
+        }
+        if (table == FT_ATOMIC_LOAD_PTR_ACQUIRE(so->table) &&
+            startkey == FT_ATOMIC_LOAD_PTR_ACQUIRE(ep->key)) {
+            assert(cmp == SET_LOOKKEY_FOUND || cmp == SET_LOOKKEY_NO_MATCH);
+            return cmp;
+        }
```
What are we checking here and why? For the original code the `PyObject_RichCompareBool` could have changed the set, and these checks guard against this. However, in the free-threading setting a concurrent thread can change the set `so` after the `table == FT_ATOMIC_LOAD_PTR_ACQUIRE(so->table) && startkey == FT_ATOMIC_LOAD_PTR_ACQUIRE(ep->key)`, but before the `return cmp;`. So any calling method cannot assume that when `SET_LOOKKEY_FOUND` is returned the `key` is still in the set.

Maybe adding the check help to preserve backwards compatibility in case of a mutating rich compare?

nascheme replied:
I believe the "why" is to handle `PyObject_RichCompareBool()` mutating the set, via the restart.  I think we want to preserve that, even for the free-threaded build, since some code that running single-threaded might assume set lookups handle mutation that way.

eendebakpt commented on Objects/setobject.c:
```diff
@@ -63,6 +63,147 @@ static PyObject _dummy_struct;
+#endif
+
+static inline Py_ALWAYS_INLINE int
+set_compare_entry(PySetObject *so, setentry *table, setentry *entry,
```
This method is only called with `so` in a critical section and I think the locking of `so` is required to make this implementation valid. Correct? If so, can we add an assert to make sure `so` is locked and/or change the name to `set_compare_entry_lock_held`?

nascheme replied:
Yeah, I think `set_compare_entry_lock_held` is a bit clearer.

nascheme committed:
Minor code cleanup based on review.

* Add comment to _PySet_Contains()
* Rename set_compare_entry() to set_compare_entry_lock_held().
* Remove unneeded `startkey == NULL` test.

nascheme committed:
Merge 'origin/main' into set_lockfree_contains

nascheme committed:
Improve unit tests.

The test_contains_frozenset() was not actually doing anything useful,
rewrite it.  Add test_contains_hash_mutate() which triggers unusual
code paths due to `__hash__` on the key mutating the container.

nascheme committed:
wip: improve test

nascheme committed:
Merge branch 'origin/main' into set_lockfree_contains
